### PR TITLE
Merge from downstream to upstream

### PR DESCRIPTION
**Pull request by melroy89**
- Introducing GitHub Actions
- Remove travis / landscape
- Introducing dedicated caching service
- Fixing user-agent page
- Making w3schools state page obsolete + removal
- Clean-up/refactoring code+ better coding standard (`black`)
- Update `requirements.txt`

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -21,3 +21,4 @@ __pycache__
 *.swo
 *~
 htmlcov
+python-env
```
Comment (cclauss) [inline .gitignore]: ⛔ in https://github.com/hellysmile/fake-useragent/pull/136/files means that GitHub UI expects text files to end in one and only one `\n`.

Comment (cclauss) [inline .gitignore]:
```suggestion
python-env
```

Comment (melroy89) [reply .gitignore]: True, however it's not the end of the world btw.

Comment (cclauss) [inline .gitignore]: ⛔ is not resolved.
Screenshot 2022-11-02 at 21 48 59

File: README.md [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+### Installation
+
+For now:
+
+```sh
+pip install .
+```
+
+_Soon again:_
+
```
Comment (cclauss) [inline README.md]: This works today.
```suggestion
```

File: README.md [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+# in case if something went wrong, one more time it is REALLY!!! rare case
+ua.random == 'your favorite Browser'
+```
+
+Want to control location of data file? (version 0.1.4 added)
+
+```py
+import fake_useragent
+
+# I am STRONGLY!!! recommend to use version suffix
```
Comment (cclauss) [inline README.md]:
```suggestion
# I am STRONGLY!!! recommend using a version suffix
```

File: README.md [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+
+Please, do not use if you don't understand why you need this.
+This is magic for rarely extreme case.
+
+### Experiencing issues???
+
+Make sure that you using latest version!!!
+
+```sh
+pip install -U fake-useragent
```
Comment (cclauss) [inline README.md]: `pip install` has [options](https://pip.pypa.io/en/stable/cli/pip_install/#options) for user, upgrade, upgrade-strategy, and use-pep517 so it is better to be explicit.
```suggestion
pip install --upgrade fake-useragent
```

File: cache_scraper/useragentscraper/middlewares.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        # Should return None or raise an exception.
+        return None
+
+    def process_spider_output(self, response, result, spider):
+        # Called with the results returned from the Spider, after
+        # it has processed the response.
+
+        # Must return an iterable of Request, or item objects.
+        for i in result:
+            yield i
```
Comment (cclauss) [inline cache_scraper/useragentscraper/middlewares.py]:
```suggestion
        yield from result
```

File: README.md (hunk 4 above)
Comment (melroy89) [reply README.md]: You have only `-U` or `--upgrade`: https://pip.pypa.io/en/stable/cli/pip_install/#cmdoption-U ??

There is no `--update`?

File: cache_scraper/useragentscraper/middlewares.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        pass
+
+    def process_start_requests(self, start_requests, spider):
+        # Called with the start requests of the spider, and works
+        # similarly to the process_spider_output() method, except
+        # that it doesn’t have a response associated.
+
+        # Must return only requests (not items).
+        for r in start_requests:
+            yield r
```
Comment (cclauss) [inline cache_scraper/useragentscraper/middlewares.py]:
```suggestion
        yield from start_requests
```

File: cache_scraper/useragentscraper/middlewares.py (hunk 5 above)
Comment (melroy89) [reply cache_scraper/useragentscraper/middlewares.py]: This is generated template code from scrapy. I will leave it like this if you don't mind.

File: cache_scraper/useragentscraper/middlewares.py (hunk 6 above)
Comment (danger89) [reply cache_scraper/useragentscraper/middlewares.py:53]: This is generated template code from scrapy. I will leave it like this if you don't mind.

File: cache_scraper/useragentscraper/middlewares.py (hunk 5 above)
Comment (cclauss) [reply cache_scraper/useragentscraper/middlewares.py]: Since Python 3.3

Comment (melroy89) [reply cache_scraper/useragentscraper/middlewares.py]: OK I will apply the improvement.

File: README.md (hunk 4 above)
Comment (cclauss) [reply README.md]: Thx!  Fixed inline.

**Comment by danger89** [comment]
So now I only need to debug this issue: https://github.com/danger89/fake-useragent/actions/runs/3381208500/jobs/5614863529

**Comment by melroy89** [comment]
Build is green (no tests are disabled!). See: https://github.com/danger89/fake-useragent/actions/runs/3381385589

image

Will make it soon be 100% code coverage.

**Comment by adamahlstrom4321** [comment]
Hi!

Will 0.1.13 be downloadable through pip?

**Comment by melroy89** [comment]
> Hi!
> 
> Will 0.1.13 be downloadable through pip?

Yes it will! We are migrating to a separate GitHub organization. After which I will add a deployment job towards PyPi 🤟🏽

**Comment by melroy89** [comment]
Version 0.1.13 is deployed NOW to PyPi (which you can install using `pip`):

```sh
pip install fake-useragent
```

New GitHub repo location: https://github.com/fake-useragent/fake-useragent

Deployment CI/CD:
image

**Pull request merged**